Feature: Header Component

filiphosko opened the pull request:
This is the first version of header/navbar element. I need feedback about what to change/improve. There are still things missing (tests, proptypes validation), I will add them later after reviews and tweaks. This PR targets the issue https://github.com/react-native-training/react-native-elements/issues/47.

Monte9 commented:
@filiphosko This is a great start! Thanks for owning this roadmap feature. 💯 

Here's a list of all the things that need to be addressed:
1. Add screenshots for all possible combinations of props. This helps improve the component from a visual standpoint as we can ensure we are making it as customizable as possible with clean UI. 
2. Refactor to deconstruct props within the class/object. Look at this PR #285 to see what I am talking about. 
3. Add 1 simple test case. Just a simple passing test is a good start too. Look at this PR if you need a reference #286 
4. Add barebones docs (look in the `docs/` folder)

filiphosko commented:
I refactored the component to support passing in components through props and also added some simple (and superficial) tests. I don't know why the Codecov checks are not passing tho (all tests were passing locally).

I will add some documentation and screenshots next.

filiphosko commented:
I'm finishing the docs, I have to make some screenshots and push the changes here.

binoy14 commented:
@filiphosko This looks amazing! Will review it soon. Thanks for the PR

filiphosko commented:
Thanks, I will probably revert the 'unification' of styling that I did with the second last commit (basically there is a top offset by -20 on Android which pushes the header under the StatusBar and makes it look like on iOS). Now I don't see it as a good idea. Apart from that everything else was meant to be ;) I'm open to improvements.

binoy14 commented on src/header/Header.js:
```diff
@@ -0,0 +1,129 @@
+import isEmpty from 'lodash.isempty';
+import Icon from '../icons/Icon';
+
+export const DummyNavButton = () => (
```
Is this temporary? P.s I like the name

Monte9 replied:
can we rename to `DefaultNavButton` if that makes sense in this context? 
I'll leave it up to you to the decide.

binoy14 commented on src/header/Header.js:
```diff
@@ -0,0 +1,129 @@
+  return childrenArray;
+}
+
+const Header = (props) => {
```
I am thinking let's split the components in seperate files. That way we can maintain it easily as well as test them easily. What do you think?

filiphosko replied:
Sure, we can. In my first interation I had them split in separate files, then I changed my mind since they were very small and I could avoid unnecessary imports. Sometimes I try to balance it (like when components are really tiny), but maybe it would be better also in this case to have them separate.

binoy14 commented on src/header/Header.js:
```diff
@@ -0,0 +1,129 @@
+  <View style={styles.dummyNavBtn} />
+);
+
+export const Title = (props) => {
```
This is just a text component right? Why export it?

filiphosko replied:
I exported it only to be able to test it (to have 100% test coverage).

binoy14 commented on src/header/Header.js:
```diff
@@ -0,0 +1,129 @@
+  } = props;
+
+  return (
+    <Text
```
Let's use text from react native elements as it handles font size based on pixel ratio

filiphosko replied:
Ok, I'll swap it 👍

binoy14 commented on src/header/Header.js:
```diff
@@ -0,0 +1,129 @@
+  );
+};
+
+export const NavButton = (props) => {
```
Again it's just the icon component why export it?

filiphosko replied:
To be able to test it, would I be able to do it otherwise?

Monte9 commented on src/header/Header.js:
```diff
@@ -0,0 +1,129 @@
+};
+
+const styles = StyleSheet.create({
+  dummyNavBtn: {
```
rename to `defaultNavButton`?

filiphosko replied:
Yep, we can, but semantically it's just a placeholder and not even a button, that's why I named it 'dummy'. But we can rename it for sure.

Monte9 commented on src/header/__tests__/Header.js:
```diff
@@ -0,0 +1,170 @@
+    expect(toJson(component)).toMatchSnapshot();
+  });
+
+  it('should accept props', () => {
```
tests are great. 💯 gj!

filiphosko replied:
Thanks, I tried to keep codecov satisfied.

binoy14 commented:
@filiphosko Is this ready to be merged? I personally feel that we should split small components in separate files. Seems like you agree too.

cc @Monte9

filiphosko commented:
@binoy14 Yep, I agree, but sometimes I hesitate to do it since I try to avoid unnecessary import where I can. If I would split them, 1) they would need to be imported into Header.js, 2) each of them will need to import React

filiphosko commented:
Sorry, I didn't have time to do it, so hopefully it's ok now.

binoy14 commented:
Thanks @filiphosko
